gpr.py

```diff
@@ -1,5 +1,62 @@
+import matplotlib
+matplotlib.use('agg')
+from matplotlib import cm, pyplot
+import numpy
+from sklearn import gaussian_process
+import io
+import arff
+import os
+import json
+
+
+def branin_mesh(X0, X1):
+    # b,c,t = 5.1/(4.*(pi)**2), 5./pi, 1./(8.*pi)
+    b, c, t = 0.12918450914398066, 1.5915494309189535, 0.039788735772973836
+    u = X1 - b*X0**2 + c*X0 - 6
+    r = 10.*(1. - t) * numpy.cos(X0) + 10
+    Z = u**2 + r
+    return Z
+
+
 def run_gpr():
-    print("hello world")
+    npoints = 15
+    X0_vec = numpy.linspace(-5., 10., npoints)
+    X1_vec = numpy.linspace(0., 15., npoints)
+    X0, X1 = numpy.meshgrid(X0_vec, X1_vec)
+    Z = branin_mesh(X0, X1)
+
+    dids = json.loads(os.getenv('DIDS', None))
+
+    if not dids:
+        print("no dids")
+        return
+
+    for did in dids:
+        # get the ddo from disk
+        filename = '/data/ddos/' + did
+        with open(filename) as datafile:
+            res = arff.load(datafile)
+
+    mat = numpy.stack(res["data"])
+    [X, y] = numpy.split(mat, [2], axis=1)
+
+    model = gaussian_process.GaussianProcessRegressor()
+    model.fit(X, y)
+    yhat = model.predict(X, return_std=False)
+    Zhat = numpy.reshape(yhat, (npoints, npoints))
+
+    # plot data + model
+    fig, ax = pyplot.subplots(subplot_kw={"projection": "3d"})
+    ax.plot_wireframe(X0, X1, Z, linewidth=1)
+    ax.scatter(X0, X1, Zhat, c="r", label="model")
+    pyplot.title("Data + model")
+
+    buf = io.BytesIO()
+    pyplot.savefig(buf, format='png')
+    buf.seek(0)
+    print(buf.read())
+    buf.close()
```

Comment:
I know in slack I'd said "ok" to just outputting an image. But thinking more about it, this is not a good example for c2d. Much better is to output the gpr model itself.

Here's how: https://scikit-learn.org/stable/modules/model_persistence.html

I do suggest to keep the plot-creation around here because it helps debugging. It it into an "if False" code block. And I suggest to delete saving the plot.

Reply:
Ok, so I just talked to Alex and he said it's possible to create both files as results. I've been operating under the assumption that I can only print stuff to the log, but in fact I can have concrete result files :) So I'll switch to that and save both.

Reply:
This is also resolvable now. I am pickling the model and saving it as a result. If we need to implement json instead, let me know, but I believe pickling is more integrated with the flow, e.g. the data scientist will use c2d and stay in the same environment to run it, so pickle is appropriate. I am open to anything, though.